### PR TITLE
Separate data storage abstraction for MergeTree

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
.

### CODE REVIEW

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp [hunk 1]
```diff
@@ -98,27 +98,28 @@ void IMergeTreeDataPart::MinMaxIndex::load(const MergeTreeData & data, const Par
...
-    const String & part_path,
+    const DataPartStorageBuilderPtr & data_part_storage_builder,
     Checksums & out_checksums) const
 {
     if (!initialized)
-        throw Exception("Attempt to store uninitialized MinMax index for part " + part_path + ". This is a bug.",
-            ErrorCodes::LOGICAL_ERROR);
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Attempt to store uninitialized MinMax index for part {}. This is a bug.",
```
Comment (amosbird) [inline src/Storages/MergeTree/IMergeTreeDataPart.cpp:121]: trailing period.

File: src/Storages/MergeTree/DataPartStorageOnDisk.cpp [hunk 2]
```diff
@@ -0,0 +1,877 @@
...
+        disk->removeSharedRecursive(fs::path(dir) / "", !can_remove_shared_data, names_not_to_remove);
+    }
+}
+
+DataPartStorageOnDisk::DisksSet::const_iterator DataPartStorageOnDisk::isStoredOnDisk(const DisksSet & disks) const
+{
+    return disks.find(volume->getDisk());
+}
+
+DataPartStoragePtr DataPartStorageOnDisk::getProjection(const std::string & name) const
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Storages/MergeTree/DataPartStorageOnDisk.cpp:294]: I just think it is more convenient. Projection has the same structure in internal folder, and interface is generally the same

File: src/Storages/MergeTree/DataPartStorageOnDisk.cpp [hunk 3]
```diff
@@ -0,0 +1,877 @@
...
+}
+
+void DataPartStorageOnDisk::writeColumns(const NamesAndTypesList & columns, const WriteSettings & settings) const
+{
+    std::string path = fs::path(root_path) / part_dir / "columns.txt";
+
+    try
+    {
+        {
+            auto buf = volume->getDisk()->writeFile(path + ".tmp", 4096, WriteMode::Rewrite, settings);
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Storages/MergeTree/DataPartStorageOnDisk.cpp:382]: Agree

File: src/Storages/MergeTree/MergeTreeDataPartWriterOnDisk.h [hunk 4]
```diff
@@ -128,7 +129,7 @@ class MergeTreeDataPartWriterOnDisk : public IMergeTreeDataPartWriter
 
     const MergeTreeIndices skip_indices;
 
-    const String part_path;
+    //const String part_path;
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeDataPartWriterOnDisk.h:132]: Remove

File: src/Storages/MergeTree/MergeTreeIndices.h [hunk 5]
```diff
@@ -23,6 +24,9 @@ namespace ErrorCodes
     extern const int NOT_IMPLEMENTED;
 }
 
+// class IDataPartStorage;
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeIndices.h:27]: Remove

File: src/Storages/MergeTree/MutateTask.cpp [hunk 6]
```diff
@@ -653,10 +654,11 @@ struct MutationContext
     MutationsInterpreter::MutationKind::MutationKindEnum mutation_kind
         = MutationsInterpreter::MutationKind::MutationKindEnum::MUTATE_UNKNOWN;
 
-    VolumePtr single_disk_volume;
+    //VolumePtr single_disk_volume;
```
Comment (alesapin) [inline src/Storages/MergeTree/MutateTask.cpp:657]: Remove

File: src/Storages/MergeTree/MergeTreeDataPartCompact.cpp [hunk 7]
```diff
@@ -137,7 +139,7 @@ bool MergeTreeDataPartCompact::hasColumnFiles(const NameAndTypePair & column) co
 void MergeTreeDataPartCompact::checkConsistency(bool require_part_metadata) const
 {
     checkConsistencyBase();
-    String path = getFullRelativePath();
+    // String path = getRelativePath();
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeDataPartCompact.cpp:142]: Remove

File: src/Storages/StorageMergeTree.cpp [hunk 8]
```diff
@@ -1739,29 +1739,25 @@ CheckResults StorageMergeTree::checkData(const ASTPtr & query, ContextPtr local_
 
     for (auto & part : data_parts)
     {
-        auto disk = part->volume->getDisk();
-        String part_path = part->getFullRelativePath();
+        //auto disk = part->volume->getDisk();
```
Comment (alesapin) [inline src/Storages/StorageMergeTree.cpp:1742]: Remove

File: src/Storages/MergeTree/MutateTask.cpp [hunk 9]
```diff
@@ -1479,8 +1497,8 @@ bool MutateTask::prepare()
     ctx->new_data_part->setSerializationInfos(new_infos);
     ctx->new_data_part->partition.assign(ctx->source_part->partition);
 
-    ctx->disk = ctx->new_data_part->volume->getDisk();
-    ctx->new_part_tmp_path = ctx->new_data_part->getFullRelativePath();
+    // ctx->disk = ctx->new_data_part->volume->getDisk();
```
Comment (alesapin) [inline src/Storages/MergeTree/MutateTask.cpp:1500]: Remove

File: src/Storages/MergeTree/MutateTask.cpp [hunk 10]
```diff
@@ -1203,12 +1207,12 @@ class MutateSomePartColumnsTask : public IExecutableTask
...
         NameSet hardlinked_files;
         /// Create hardlinks for unchanged files
-        for (auto it = ctx->disk->iterateDirectory(ctx->source_part->getFullRelativePath()); it->isValid(); it->next())
+        for (auto it = ctx->source_part->data_part_storage->iterate(); it->isValid(); it->next())
         {
             if (ctx->files_to_skip.contains(it->name()))
                 continue;
 
-            String destination = ctx->new_part_tmp_path;
+            String destination; // = ctx->new_part_tmp_path;
```
Comment (alesapin) [inline src/Storages/MergeTree/MutateTask.cpp:1215]:
```suggestion
            String destination;
```

**Review by alesapin — APPROVED** [review]

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp [hunk 11]
```diff
@@ -615,26 +598,26 @@ String IMergeTreeDataPart::getColumnNameWithMinimumCompressedSize(
...
+        throw Exception("Could not find a column of minimum size in MergeTree, part " + data_part_storage->getFullPath(), ErrorCodes::LOGICAL_ERROR);
 
     return *minimum_size_column;
 }
 
-String IMergeTreeDataPart::getFullPath() const
-{
-    if (relative_path.empty())
-        throw Exception("Part relative_path cannot be empty. It's bug.", ErrorCodes::LOGICAL_ERROR);
+// String IMergeTreeDataPart::getFullPath() const
```
Comment (tavplubix) [inline src/Storages/MergeTree/IMergeTreeDataPart.cpp:606]: Please remove commented code

**Pull request merged**